File: istaroth/agd/renderables/weapon.py
"""Weapon story processing and rendering."""

from istaroth import utils
from istaroth.agd import (
    id_types,
    processed_types,
    repo,
)
from istaroth.agd.renderables import readable as _readable
from istaroth.text import types as text_types


def get_weapon_info(
    weapon_id: str, *, data_repo: repo.DataRepo
) -> processed_types.WeaponInfo | None:
    """Assemble a weapon's story document from its authoritative weapon config.

    Follows weapon storyId -> DocumentExcelConfigData -> ordered page localization
    ids -> readable files, joining the pages into one document. Returns None when
    the weapon has no story (storyId 0, no document, or no page has on-disk
    content), mirroring the artifact-set discovery model. Reading each page also
    marks it accessed, keeping rendered pages out of the generic Readables
    catch-all; the unrendered base/placeholder files it leaves behind are dropped
    there by the empty/placeholder content skip.
    """
    text_map = data_repo.build_text_map_tracker()
    readables = data_repo.build_readables_tracker()

    if (
        weapon := data_repo.load_weapon_excel_config_data().get(int(weapon_id))
    ) is None:
        raise ValueError(f"Weapon configuration not found for weapon ID: {weapon_id}")

    if (story_id := weapon["storyId"]) == 0:
        return None

    if (doc_item := data_repo.load_document_excel_config_data().get(story_id)) is None:
        return None

    ordered_loc_ids = dict.fromkeys(
        doc_item["questContentLocalizedId"]
        + doc_item["questIDList"]
        + doc_item.get("CUSTOM_addlLocalID", [])
    )
    readable_paths = data_repo.build_localization_id_to_readable_path_mapping()
    story_pages = [
        content
        for loc_id in ordered_loc_ids
        if (path := readable_paths.get(loc_id)) is not None
        and (content := readables.get_content(path))
    ]
    if not story_pages:
        return None

    if (name := text_map.get_optional(weapon["nameTextMapHash"])) is None:
        raise ValueError(f"Missing name for weapon ID {weapon_id}")

    return processed_types.WeaponInfo(
        weapon_id=weapon_id,
        name=name,
        description=text_map.get(weapon["descTextMapHash"], ""),
        story_pages=story_pages,
    )


def render_weapon(
    weapon_info: processed_types.WeaponInfo,
) -> processed_types.RenderedItem:
    """Render a weapon's assembled story document into RAG-suitable format."""
    safe_name = utils.make_safe_filename_part(weapon_info.name)
    filename = f"{weapon_info.weapon_id}_{safe_name}.txt"

    content_lines = [f"# {weapon_info.name}\n"]
    if weapon_info.description:
        content_lines.append(f"{weapon_info.description}\n")
    content_lines.append("\n\n---\n\n".join(weapon_info.story_pages))

    return processed_types.RenderedItem(
        text_metadata=text_types.TextMetadata(
            category=text_types.TextCategory.AGD_WEAPON,
            title=weapon_info.name,
            id=int(weapon_info.weapon_id),
            relative_path=f"{text_types.TextCategory.AGD_WEAPON.value}/{filename}",
        ),
        content="\n".join(content_lines),
    )

File: rust/istaroth-agd-regen/src/renderables/weapon.rs
//! Port of renderables/weapon.py.

use crate::firstseen::Domain;
use crate::issues::Scope;
use crate::rendered_item::RenderedItem;
use crate::repo::Repo;
use crate::util;
use crate::vh::{ValueExt, int_array};
use anyhow::{Context, Result, anyhow};
use rustc_hash::FxHashSet;

/// (CHS, non-CHS) per-pass error limits (see e.g. `artifact::ERROR_LIMITS`).
pub const ERROR_LIMITS: (usize, usize) = (50, 200);

/// Weapons pass discovery: sorted weapon ids.
pub fn discover(repo: &Repo) -> Result<Vec<i64>> {
    let mut ids: Vec<i64> = repo.excel.weapon.keys().copied().collect();
    ids.sort();
    Ok(ids)
}

/// Assemble a weapon's story document from its authoritative weapon config.
///
/// Follows weapon storyId -> DocumentExcelConfigData -> ordered page
/// localization ids -> readable files, joining the pages into one document.
/// Returns None when the weapon has no story (storyId 0, no document, or no
/// page has on-disk content), mirroring the artifact-set discovery model.
/// Reading each page also marks it accessed, keeping rendered pages out of
/// the generic Readables catch-all; the unrendered base/placeholder files it
/// leaves behind are dropped there by the empty/placeholder content skip.
pub fn process(repo: &Repo, scope: &Scope, weapon_id: i64) -> Result<Option<RenderedItem>> {
    let weapon = repo
        .excel
        .weapon
        .get(&weapon_id)
        .ok_or_else(|| anyhow!("Weapon configuration not found for weapon ID: {weapon_id}"))?;
    let story_id = weapon.i("storyId")?;
    if story_id == 0 {
        return Ok(None);
    }
    let Some(doc_item) = repo.excel.document.get(&story_id) else {
        return Ok(None);
    };
    // Ordered dedup of questContentLocalizedId + questIDList +
    // CUSTOM_addlLocalID: first occurrence keeps its position.
    let mut loc_ids: Vec<i64> = int_array(doc_item.f("questContentLocalizedId")?)?;
    loc_ids.extend(int_array(doc_item.f("questIDList")?)?);
    if let Some(addl) = doc_item.get("CUSTOM_addlLocalID") {
        loc_ids.extend(int_array(addl).context("CUSTOM_addlLocalID")?);
    }
    let mut seen: FxHashSet<i64> = FxHashSet::default();
    loc_ids.retain(|id| seen.insert(*id));
    let mut story_pages: Vec<String> = Vec::new();
    for loc_id in &loc_ids {
        if let Some(filename) = repo.loc_id_to_readable_filename.get(loc_id)
            && let Some(content) = repo.readable_content(filename, scope)
            && !content.is_empty()
        {
            story_pages.push(content.to_string());
        }
    }
    if story_pages.is_empty() {
        return Ok(None);
    }

    let name = repo.tm.get_required(weapon.i("nameTextMapHash")?, scope)?;
    let description = repo.tm.get_or(weapon.i("descTextMapHash")?, "", scope)?;

    let safe_name = util::make_safe_filename_part(&name);
    let mut content_lines = vec![format!("# {name}\n")];
    if !description.is_empty() {
        content_lines.push(format!("{description}\n"));
    }
    content_lines.push(story_pages.join("\n\n---\n\n"));
    let versions = repo.first_seen.resolve_int(Domain::Weapon, weapon_id)?;
    Ok(Some(RenderedItem::new(
        "agd_weapon",
        name,
        weapon_id,
        format!("{weapon_id}_{safe_name}.txt"),
        versions,
        content_lines.join("\n"),
    )))
}
